packages/flutter/lib/src/gestures/tap.dart

```diff
@@ -98,6 +98,197 @@ typedef GestureTapCallback = void Function();
+/// want to build a custom tap recognizer, extend this class by overriding
+/// [isPointerAllowed] and the handler methods.
+///
+/// See also:
+///
+///  * [TapGestureRecognizer], a ready-to-use tap recognizer that recognizes
+///    taps of the primary button and taps of the secondary button.
+///  * [ModalBarrier], a widget that uses a custom tap recognizer that accepts
+///    any buttons.
+abstract class BaseTapGestureRecognizer extends PrimaryPointerGestureRecognizer {
+  /// Creates a tap gesture recognizer.
+  BaseTapGestureRecognizer({ Object debugOwner, Duration deadline = kPressTimeout })
+    : super(deadline: deadline, debugOwner: debugOwner);
+
+  bool _sentTapDown = false;
+  bool _wonArenaForPrimaryPointer = false;
+
+  PointerDownEvent _down;
+  PointerUpEvent _up;
+
+  /// A pointer that might cause a tap has contacted the screen.
+  ///
+  /// This triggers after the down event, once a short timeout ([deadline]) has
+  /// elapsed, or once the gestures has won the arena, whichever comes first.
+  ///
+  /// The parameter `down` is the down event of the primary pointer that started
+  /// the tap sequence.
+  ///
+  /// If the gesture doesn't win the arena, [handleTapUp] is called next.
+  /// Otherwise, [handleTapCancel] is called next.
+  @protected
+  void handleTapDown({ PointerDownEvent down });
+
+  /// A pointer that will trigger a tap has stopped contacting the screen.
+  ///
+  /// This triggers on the up event, if the gesture wins the arena with it
+  /// or has previously won.
+  ///
+  /// The parameter `down` is the down event of the primary pointer that started
+  /// the tap sequence, and `up` is the up event that ended the tap sequence.
+  ///
+  /// If the gesture doesn't win the arena, [handleTapCancel] is called
+  /// instead.
+  @protected
+  void handleTapUp({ PointerDownEvent down, PointerUpEvent up });
+
+  /// The pointer that previously triggered [handleTapDown] will not end up
+  /// causing a tap.
+  ///
+  /// This triggers once the gesture loses the arena, if [handleTapDown] has
+  /// been previously triggered.
+  ///
+  /// The parameter `down` is the down event of the primary pointer that started
+  /// the tap sequence; `cancel` is the cancel event, which might be null;
+  /// `reason` is a short description of the cause if `cancel` is null, which
+  /// can be `"forced"` if other gestures won the arena, or `"spontaneous"`
+  /// otherwise.
+  ///
+  /// If the gesture wins the arena, [onSecondaryTapUp] is called instead.
+  @protected
+  void handleTapCancel({ PointerDownEvent down, PointerCancelEvent cancel, String reason });
+
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    if (state == GestureRecognizerState.ready) {
+      // `_down` must be assigned in this method instead of `handlePrimaryPointer`,
+      // because `acceptGesture` might be called before `handlePrimaryPointer`,
+      // which relies on `_down` to call `_delegate.onDown`.
```

Comment:
What's the _delegate.onDown this is talking about?

Reply:
It was the result of a previous design... Thanks for spotting this (and all others)!